contents/security/sqlInjection.md

```diff
@@ -0,0 +1,216 @@
+
+The attacker will know that there is no table named "Users" in the database. If, after a number 
+of tries, a different message appears:
+
+```html
+ORA-01789: query block has incorrect number of result columns.
+```
+
+Then the attacker would have found a table which exists in the database, as the error message 
+has changed. Now that an existing table has been found, the attacker can refine the query further. 
+In this example, the attack failed because the number of columns was inconsistent with the expected 
+result. The attacker may try "SELECT Username, Password" next, which will limit the retrieved data 
+to be two columns, one for each parameter requested.
+
+
+### Blind SQL Injection ###
+
+Even when there is no detailed error from the database displayed, attackers can gather 
+information though other means, such as by testing different true or false statements.
+
+When attempting to login to a website normally, the attacker may see this error message:
+
+```html
+Error: Invalid credentials!
+```
+
+The message does not give enough information to the attacker. It is unknown whether 
+the username or password is wrong here. However, this is not enough to deter a SQL 
+injection attack! If the following attack is attempted and successful:
+
+```html
+Username: foo
+Password: bar’ OR ‘’=’
+```
+
+Then the attacker knows that the form is vulnerable to following attacks. Once a SQL 
+injection vulnerability is discovered, the attacker can continue to exploit the 
+database as per normal SQL Injection. For example, if the attacker knew of a table 
+in the database named 'Users', the attacker can enter this malicious string, which 
+would delete the entire table and its associated data:
+
+```html
+Username: foo
+Password: bar’; DROP TABLE Users;--
+```
+
+## Preventing SQL Injection ##
+
+There are a couple of ways to protect your website against SQL injection attacks. The two most
+common ways are:
+
+### Using Prepared Statements using Parameterized Queries ###
+
+By defining all the SQL code first and passing in the parameters afterwards, you can make 
+the database distinguish the difference between code and data. It would treat the values 
+entered by the user as a parameter, and would not allow it to alter the query being 
+executed. The way to achieve this varies by language, but it is easy to implement and effective.
+
+For example, instead of writing this in Java:
+
+```html
+String query = "SELECT account_balance FROM user_data WHERE user_name = " + request.getParameter("customerName");
+Statement statement = connection.createStatement();
+ResultSet results = statement.executeQuery( query );
+```
+
+You can prepare the statement like this:
+
+```html
+ String custname = request.getParameter("customerName");
+ String query = "SELECT account_balance FROM user_data WHERE user_name = ? ";
+ PreparedStatement pstmt = connection.prepareStatement( query );
+ pstmt.setString( 1, custname ); 
+ ResultSet results = pstmt.executeQuery( );
```

Comment:
You shoukd explain a bit more about the meaning of "?" and its connection to the next statement